src/MSBuild/TerminalLogger/TerminalLogger.cs

```diff
@@ -609,6 +630,32 @@ private void TargetFinished(object sender, TargetFinishedEventArgs e)
     {
     }
 
+    private void TryReadSourceControlInformationForProject(BuildEventContext? context, IEnumerable<ITaskItem>? sourceRoots)
+    {
+        if (context is null || sourceRoots is null)
+        {
+            return;
+        }
+
+        var projectContext = new ProjectContext(context);
+        if (_projects.TryGetValue(projectContext, out Project? project))
+        {
+            if (project.SourceRoot is not null)
+            {
+                return;
+            }
+            var sourceControlSourceRoot = sourceRoots.FirstOrDefault(root => !string.IsNullOrEmpty(root.GetMetadata("SourceControl")));
+            if (sourceControlSourceRoot is not null)
+            {
+                // This takes the first root from source control the first time it's added to the build.
+                // This seems to be the Target InitializeSourceControlInformationFromSourceControlManager.
+                // So far this has been acceptable, but if a SourceRoot would be modified by a task later on
+                // (e.g. TranslateGitHubUrlsInSourceControlInformation) we would lose that modification.
```

Comment:
In this case it's not an `Update` unfortunately - the targets explicitly remove all of the previous items and re-add the new items. 
I think it would be useful separately to have an `UpdateItem` but it wouldn't help me here :)